include/aws/crt/ImdsClient.h

```diff
@@ -0,0 +1,441 @@
+        {
+
+            struct AWS_CRT_CPP_API ImdsClientConfig
+            {
+                ImdsClientConfig() : Bootstrap(nullptr) {}
+
+                /**
+                 * Connection bootstrap to use to create the http connection required to
+                 * query resource from the Ec2 instance metadata service
+                 */
+                Io::ClientBootstrap *Bootstrap;
+
+                /* Should add retry strategy support once that is available */
+            };
+
+            /**
+             * https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
+             */
+            struct AWS_CRT_CPP_API IamProfile
+            {
+                DateTime lastUpdated;
+                StringView instanceProfileArn;
+                StringView instanceProfileId;
+            };
+
+            /**
+             * A convenient class for you to persist data from IamProfile, which has StringView members.
+             */
+            struct AWS_CRT_CPP_API IamProfileStore
+            {
+                IamProfileStore() {}
+                IamProfileStore(const IamProfile &other)
+                    : lastUpdated(other.lastUpdated),
+                      instanceProfileArn(other.instanceProfileArn.data(), other.instanceProfileArn.size()),
+                      instanceProfileId(other.instanceProfileId.data(), other.instanceProfileId.size())
+                {
+                }
+
+                IamProfileStore &operator=(const IamProfile &other)
+                {
+                    lastUpdated = other.lastUpdated;
+                    instanceProfileArn = String(other.instanceProfileArn.data(), other.instanceProfileArn.size());
+                    instanceProfileId = String(other.instanceProfileId.data(), other.instanceProfileId.size());
+                    return *this;
+                }
+
+                DateTime lastUpdated;
+                String instanceProfileArn;
+                String instanceProfileId;
+            };
+
+            /**
+             * Block of per-instance EC2-specific data
+             *
+             * https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-identity-documents.html
+             */
+            struct AWS_CRT_CPP_API InstanceInfo
+            {
+                /* an array of StringView */
+                Vector<StringView> marketplaceProductCodes;
+                StringView availabilityZone;
+                StringView privateIp;
+                StringView version;
+                StringView instanceId;
+                /* an array of StringView */
+                Vector<StringView> billingProducts;
+                StringView instanceType;
+                StringView accountId;
+                StringView imageId;
+                DateTime pendingTime;
+                StringView architecture;
+                StringView kernelId;
+                StringView ramdiskId;
+                StringView region;
+            };
```

Comment:
Whoa, that's a lot of non-owning members!

Maybe we should rename this `InstanceInfoView` to hint that it doesn't own its memory?

Also, if the user wanted to persist this object, they'd need to declare their own type with 10+ String members and copy all the values over. We could provide that classe for them?

Or we could just have this use `String` for its members instead of `StringView`. It's not as efficient, but unless we expect this to be used in a very high-performance way, simplicity might win out.

Reply:
addressed this issue